imports/wasi_snapshot_preview1/fs.go

```diff
@@ -780,15 +778,19 @@ const direntSize = uint32(24)
 //
 // See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#fd_readdir
 // See https://github.com/WebAssembly/wasi-libc/blob/659ff414560721b1660a19685110e484a081c3d4/libc-bottom-half/cloudlibc/src/libc/dirent/readdir.c#L44
-func maxDirents(entries []fs.DirEntry, bufLen uint32) (bufused, direntCount, truncatedEntryLen uint32) {
+func maxDirents(entries []fs.DirEntry, bufLen uint32) (bufused, direntCount uint32, writeTruncatedEntry bool) {
 	lenRemaining := bufLen
 	for _, e := range entries {
 		nameLen := uint32(len(e.Name()))
 		entryLen := direntSize + nameLen
 
-		if lenRemaining == 0 {
-			// We've hit exactly the end of bufLen, so cannot write the next
+		if lenRemaining < direntSize {
```

Comment:
If we do not have enough space for a new dirent header, we want to break out of the loop and say that we have exhausted the whole buffer, not just when we have exactly 0 bytes left in the buffer.